shard/committee/assignment.go

```diff
@@ -116,62 +125,148 @@ func preStakingEnabledCommittee(s shardingconfig.Instance) shard.State {
 ) (shard.State, error) {
 	// TODO Nervous about this because overtime the list will become quite large
 	candidates := stakerReader.ValidatorCandidates()
-	stakers := make([]*staking.Validator, len(candidates))
+	essentials := map[common.Address]effective.SlotOrder{}
+	slotUsage := map[common.Address]int{}
+
+	// TODO benchmark difference if went with data structure that sorts on insert
 	for i := range candidates {
 		// TODO Should be using .ValidatorStakingWithDelegation, not implemented yet
 		validator, err := stakerReader.ValidatorInformation(candidates[i])
 		if err != nil {
 			return nil, err
 		}
-		stakers[i] = validator
+		essentials[validator.Address] = effective.SlotOrder{
+			validator.Stake,
+			validator.SlotPubKeys,
+		}
+		slotUsage[validator.Address] = len(validator.SlotPubKeys)
 	}
 
-	sort.SliceStable(
-		stakers,
-		func(i, j int) bool { return stakers[i].Stake.Cmp(stakers[j].Stake) >= 0 },
-	)
-	const sCount = 401
-	top := stakers[:sCount]
 	shardCount := int(s.NumShards())
+	maxNodePerShard := s.NumNodesPerShard()
 	superComm := make(shard.State, shardCount)
 	fillCount := make([]int, shardCount)
-	// TODO Finish this logic, not correct, need to operate EPoS on slot level,
-	// not validator level
+	hAccounts := s.HmyAccounts()
 
 	for i := 0; i < shardCount; i++ {
-		superComm[i] = shard.Committee{}
-		superComm[i].NodeList = make(shard.NodeIDList, s.NumNodesPerShard())
+		superComm[i] = shard.Committee{
+			uint32(i), make(shard.NodeIDList, s.NumNodesPerShard()),
+		}
 	}
 
-	scratchPad := &bls.PublicKey{}
-
-	for i := range top {
-		spot := int(top[i].Address.Big().Int64()) % shardCount
-		fillCount[spot]++
-		// scratchPad.DeserializeHexStr()
+	for i := range hAccounts {
+		spot := i % shardCount
+		pub := &bls.PublicKey{}
+		pub.DeserializeHexStr(hAccounts[i].BlsPublicKey)
 		pubKey := shard.BlsPublicKey{}
-		pubKey.FromLibBLSPublicKey(scratchPad)
-		superComm[spot].NodeList = append(
-			superComm[spot].NodeList,
-			shard.NodeID{
-				top[i].Address,
-				pubKey,
-				&shard.StakedMember{big.NewInt(0)},
-			},
-		)
+		pubKey.FromLibBLSPublicKey(pub)
+		superComm[spot].NodeList[fillCount[spot]] = shard.NodeID{
+			common2.ParseAddr(hAccounts[i].Address),
+			pubKey,
+			nil,
+		}
+		fillCount[spot]++
+	}
+
+	staked := effective.Apply(essentials)
+
+	sort.SliceStable(
+		staked,
+		func(i, j int) bool { return staked[i].Dec.GTE(staked[j].Dec) },
+	)
+
+	shardBig := big.NewInt(int64(shardCount))
+
+	if len(staked) <= stakedSlotsCount {
+		// WARN unlikely to happen in production but will happen as we are developing
+	}
+
+	for i := 0; i < stakedSlotsCount; i++ {
+		bucket := int(new(big.Int).Mod(staked[i].Address.Big(), shardBig).Int64())
+		slot := staked[i]
+		pubKey := essentials[slot.Address].SpreadAmong[slotUsage[slot.Address]-1]
+		slotUsage[slot.Address]--
+		// Keep going round till find an open spot
+		for j := bucket; ; j = (j + 1) % shardCount {
```

Comment:
We should just put the slot into a fixed shard based on the slot bls key. We don't need to enforce the maxNodePerShard in this release.